design/proof-erasure-coding.md

```diff
@@ -0,0 +1,419 @@
+    |///|///|///|///|///|///|///|   |    |   |   |   |   |    |
+    ---------------------------------    -------------   v    v
+
+                                <-- missing: M + 1 -->
+
+This is only (M + 1)² shards from a total of N² blocks. This gets worse when you
+go to three, four or higher dimensions. This means that our chances of detecting
+whether the data is corrupted beyond repair go down, which means that we need to
+check more shards in our Merkle storage proofs. This is exacerbated by the the
+need to counter parity blowup.
+
+##### Parity blowup ######
+
+When we perform a regular one-dimensional erasure coding, we like to use a ratio
+of 1:2 between original data (K) and total data (N), because it gives us a >50%
+chance of detecting critical data loss by checking a single shard. If we were to
+use the same K and M in a 2-dimensional setting, we'd get a ratio of 1:4 between
+original data and total data. In other words, we would blow up the original data
+by a factor of 4. This gets worse with higher dimensions.
+
+To counter this blow-up, we can choose an M that is smaller. For two dimensions,
+we could chose M = N / √2. This ensures that the total amount of data N² is
+double that of the original data K². For three dimensions we'd choose K / ∛2,
+etc. This however means that the chances of detecting critical data loss in a
+row or column go down, which means that we'd again have to sample more shards in
+our Merkle storage proofs.
+
+##### Larger encoding times #####
+
+Another drawback of multi-dimensional erasure coding is that we now need to
+erasure code the original data multiple times, and we also need to erasure code
+some of the parity data. For a two-dimensional code this means that encoding
+times go up by a factor of at least 2, and for a three-dimensional a factor of
+at least 3, etc.
+
+##### Complexity #####
+
+The final drawback of multi-dimensional erasure coding is its complexity. It is
+harder to reason about its correctness, and implementations must take great care
+to ensure that cornercases when the data is not exactly K² shaped (or K³, or...)
+are handled correctly. Decoding is also more involved because it might require
+restoring parity data before it is possible to restore the original data.
+
+##### The good news ####
+
+Despite these drawbacks, the multi-dimensional approach allows us to make the
+shards almost arbitrarily small. This allows us to compensate for the need to
+sample more shards in our Merkle proofs. For example, using a 2 dimensional
+structure of erasure coded shards in a Galois field of 2^16, we can handle 1TB
+of data with shards of size 256 bytes. When we allow parity data to take up to
+half of the total data, we would need to sample 160 shards to have a 0.999999
+chance of detecting critical data loss. This is much more than the amount of
+shards that we need in a one-dimensional setting, but the shards are much
+smaller. This leads to less hashing in a SNARK, just 40 KB.
```

Comment:
Same reasoning as above; I went with probabilities instead of incentives here.

Reply:
Note that the 40k (= 160 x 256) is just hashing the data itself. You also need the Merkle inclusion proofs of those hashes, which is an additional hashing cost. A single Merkle proof of depth 32, is 32 calls for the hash "compression function", which is about the same as hashing 32x32 = 1024 bytes of data, so that is additional 160x1024 = 160k of hashing, dominating in this case because the tiny blocks are so small.

Reply:
> additional 160x1024 = 160k of hashing

Good point, I was so used to the cost of merkle paths being negligable, that I forgot this :)
I added it to the spreadsheet and the document, thanks!